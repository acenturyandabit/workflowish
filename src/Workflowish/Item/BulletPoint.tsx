import * as React from 'react';
import { ItemTreeNode } from "~Workflowish/mvc/model";
import { ControllerActions } from "~Workflowish/mvc/controller";
import { ItemStyleParams } from ".";
import { useContextMenu } from 'react-contexify';
import { SIDECLIP_CONTEXT_MENU_ID } from '~Workflowish/Subcomponents/ContextMenu';

export const BulletPoint = (props: {
    item: ItemTreeNode,
    actions: ControllerActions,
    styleParams: ItemStyleParams,
    shouldUncollapse: boolean
}) => {
    let color = "white";
    if (props.item.searchHighlight == "SEARCH_UNCOLLAPSE") color = "orange";
    if (props.item.symlinkedNode || props.styleParams.symlinkedParent) color = "cyan";

    const hasOrSymlinkedToChildren = (props.item.children.length > 0) ||
        (props.item.symlinkedNode && props.item.symlinkedNode.children.length > 0);
    return <span style={{
        paddingLeft: hasOrSymlinkedToChildren ? "0px" : "0.2em",
        userSelect: 'none',
        cursor: "pointer",
        color
    }}
    ><span onClick={() => props.actions.getSetSelf((self: ItemTreeNode) => ({
        ...self,
        collapsed: !self.collapsed
    }))}
    >{(() => {
        let bullet = "\u25CF";
        if (props.styleParams.emptyList) bullet = ">";
        else if (hasOrSymlinkedToChildren) {
            if (props.shouldUncollapse) bullet = "\u25bc";
            else bullet = "\u25b6";
        }
        return bullet;
    })()}</span>
        {props.styleParams.showId ?
            <span style={{
                fontSize: "10px"
            }}
                onClick={(event) => {
                    const { show, hideAll } = useContextMenu({
                        id: SIDECLIP_CONTEXT_MENU_ID,
                    });
                    show({ event });
                    setTimeout(hideAll, 400);
                    navigator.clipboard.writeText(props.item.id);

                }}
            >{props.item.id}</span>
            : null}</span >;
}
