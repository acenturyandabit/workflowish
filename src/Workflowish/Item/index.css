.itemWrapperClass {
    display: flex;
    flex-direction: column;
    width: 100%;
}

.itemWrapperClass>span {
    display: inline-flex;
    width: 100%;
}