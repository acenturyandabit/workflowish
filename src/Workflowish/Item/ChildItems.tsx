import * as React from "react";
import { ItemTreeNode, TransformedDataAndSetter } from "~Workflowish/mvc/model";
import Item, { FocusActions, ItemRef, ItemStyleParams } from ".";
import { ControllerActions, makeItemActions } from "../mvc/controller"
import { FocusedActionReceiver } from "~Workflowish/mvc/focusedActionReceiver";
import { TreePath, DFSFocusManager, FocusRequest } from "~Workflowish/mvc/DFSFocus";

export const ChildItems = (props: {
    shouldUncollapse: boolean,
    item: ItemTreeNode,
    treePath: TreePath,
    focusManager: React.RefObject<DFSFocusManager>,
    styleParams: ItemStyleParams,
    itemsRefArray: React.MutableRefObject<(FocusActions | null)[]>,
    setThisAsFocused: (focusedActionReceiver: FocusedActionReceiver, focusItemKey: string) => void,
    actions: ControllerActions,
    pushRef: (id: string, ref: ItemRef) => void,
    model: TransformedDataAndSetter
}) => {
    let childrenToRender: ItemTreeNode[] = props.item.symlinkedNode ?
        props.item.symlinkedNode.children :
        props.item.children;
    if (props.styleParams.symlinkedParents.indexOf(props.item.id) != -1) {
        childrenToRender = [{
            data: "Infinite loop...",
            lastModifiedUnixMillis: 0,
            id: "",
            children: [],
            collapsed: true,
            searchHighlight: []
        }]
    }

    const itemToFocus = React.useRef<FocusRequest | undefined>();
    React.useEffect(() => {
        if (itemToFocus.current) {
            props.focusManager.current?.focusItem(itemToFocus.current);
            itemToFocus.current = undefined;
        }
    })


    return <>
        {props.shouldUncollapse ?
            <div style={{
                paddingLeft: "5px",
                borderLeft: "1px solid white",
                marginLeft: "0.5em"
            }}>
                {childrenToRender.map((item, childIdx) => {
                    const treePath = props.focusManager.current?.childPath(props.treePath, childIdx) || [];
                    return <Item
                        key={childIdx}
                        item={item}
                        styleParams={{
                            showId: props.styleParams.showId,
                            symlinkedParents: props.item.symlinkedNode ? [...props.styleParams.symlinkedParents, props.item.id] : props.styleParams.symlinkedParents
                        }}
                        setThisAsFocused={props.setThisAsFocused}
                        actions={makeItemActions({
                            thisItem: item,
                            thisPossiblySymlinkedParent: props.item,
                            treePath,
                            focusManager: props.focusManager,
                            setToFocusAfterUpdate: (focusRequest: FocusRequest) => { itemToFocus.current = focusRequest },
                            model: props.model
                        })}
                        treePath={treePath}
                        focusManager={props.focusManager}
                        model={props.model}
                        pushRef={props.pushRef}
                    ></Item>
                })}
            </div > : null
        }
    </>
};