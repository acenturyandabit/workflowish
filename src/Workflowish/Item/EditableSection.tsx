
import ContentEditable, { ContentEditableEvent } from 'react-contenteditable'
import sanitizeHtml from "sanitize-html"
import * as React from "react";
import { ItemTreeNode } from "../mvc/model";
import { ITEM_CONTEXT_MENU_ID } from "../Subcomponents/ContextMenu";
import { ControllerActions } from '~Workflowish/mvc/controller';
import { ItemStyleParams } from '.';
import { FocusedActionReceiver } from '~Workflowish/mvc/focusedActionReceiver';
import { useContextMenu } from 'react-contexify';
import { BulletPoint } from './BulletPoint';


export const EditableSection = (props: {
    item: ItemTreeNode,
    _ref: React.MutableRefObject<HTMLElement | null>,
    shouldUncollapse: boolean,
    parentActions: ControllerActions,
    styleParams: ItemStyleParams,
    focusedActionReceiver: FocusedActionReceiver,
    onFocusClick: () => void
}) => {
    let onKeyDown = (evt: React.KeyboardEvent) => {
        // TODO: inline this method
        props.focusedActionReceiver.keyCommand(evt);
    }
    if (props.styleParams.symlinkedParent) {
        onKeyDown = (evt: React.KeyboardEvent) => {
            evt.preventDefault();
            if ((evt.key == "ArrowUp" || evt.key == "ArrowDown") &&
                !(evt.altKey || evt.ctrlKey || evt.shiftKey || evt.metaKey)) {
                props.focusedActionReceiver.keyCommand({
                    key: evt.key,
                    altKey: false,
                    ctrlKey: false,
                    shiftKey: false,
                    metaKey: false,
                    preventDefault: () => {
                        // Ignore
                    }
                });
            }
        }
    }

    const onContentChange = React.useCallback((evt: ContentEditableEvent) => {
        const sanitizeConf = {
            allowedTags: ["b", "i", "a", "p"],
            allowedAttributes: { a: ["href"] }
        };
        props.parentActions.getSetSelf(oldSelf => ({
            // this needs to be a getsetter as the useEffect is run only once, and so 
            // the self item at construction time is outdated
            ...oldSelf,
            lastModifiedUnixMillis: Date.now(),
            data: sanitizeHtml(evt.currentTarget.innerHTML, sanitizeConf)
        }))
    }, [props.parentActions])

    const onContenteditableChange = (props.styleParams.symlinkedParent) ?
        (() => {/* read only*/ })
        : onContentChange;

    const memoizedInnerRef = React.useCallback(
        (contenteditableElement: HTMLElement) => {
            props._ref.current = contenteditableElement;
        }
        , []);

    const symlinkText = props.item.symlinkedNode ? <>
        &nbsp;
        <span
            onKeyDown={onKeyDown}
            onClick={props.onFocusClick}
            style={{ flex: "1 1 auto" }}
        >{props.item.symlinkedNode.data}</span>
    </> : null;

    return <span style={{ background: props.item.searchHighlight == "SEARCH_TARGET" ? "blue" : "" }}
        onContextMenu={contextEventHandler(props.parentActions)}>
        <BulletPoint
            item={props.item}
            parentActions={props.parentActions}
            styleParams={props.styleParams}
            shouldUncollapse={props.shouldUncollapse}
        ></BulletPoint>
        {/* The contentEditable needs to persist regardless of whether it is a symlink in order for focus to work correctly */}
        <ContentEditable
            innerRef={memoizedInnerRef}
            html={props.item.data}
            onChange={onContenteditableChange}
            onKeyDown={onKeyDown}
            onClick={props.onFocusClick}
            style={props.item.symlinkedNode ? { flex: "0 0 auto" } : { flex: "1 1 auto" }}
        ></ContentEditable>
        {symlinkText}
    </span>
}


const contextEventHandler = (parentActions: ControllerActions): React.MouseEventHandler<HTMLDivElement> => {
    const { show } = useContextMenu({
        id: ITEM_CONTEXT_MENU_ID,
    });
    return (event) => {
        event.preventDefault = () => {
            // prevent default context menu from being hidden
        }
        parentActions.getSetSiblingArray((siblings: ItemTreeNode[]) => {
            show({ event, props: siblings })
            return siblings;
        })
    }
}
