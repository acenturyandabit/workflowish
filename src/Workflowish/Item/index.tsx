import * as React from "react";
import { ControllerActions } from "../mvc/controller";
import { FocusedActionReceiver, makeFocusedActionReceiver } from "../mvc/focusedActionReceiver";
import { ItemTreeNode, TransformedDataAndSetter } from "../mvc/model";
import { EditableSection } from "./EditableSection";
import "./index.css"
import { ChildItems } from "./ChildItems";
import { SIDECLIP_CONTEXT_MENU_ID } from '~Workflowish/Subcomponents/ContextMenu';
import { TriggerEvent, useContextMenu } from 'react-contexify';
import { DFSFocusManager, IdAndFocusPath, TreePath } from "~Workflowish/mvc/DFSFocus";


export type FocusActions = {
    triggerFocusFromAbove: () => void;
    triggerFocusFromBelow: () => void;
    focusThis: () => void;
    scrollThisIntoView: () => void;
    focusThisEnd: () => void;
    focusRecentlyIndentedItem: () => void;
}

export type ItemStyleParams = {
    emptyList?: boolean,
    showId: boolean,
    symlinkedParents: string[]
}

const shouldBeUncollapsed = (item: ItemTreeNode): boolean => !item.collapsed || item.searchHighlight.includes("SEARCH_UNCOLLAPSE");

const Item = (props: {
    styleParams: ItemStyleParams,
    item: ItemTreeNode,
    treePath: TreePath,
    focusManager: React.RefObject<DFSFocusManager>,
    actions: ControllerActions,
    model: TransformedDataAndSetter,
    pushRef: (id: string, ref: ItemRef) => void
    setThisAsFocused: (focusedActionReceiver: FocusedActionReceiver, focusItemKey: IdAndFocusPath) => void
}) => {
    const itemsRefArray = React.useRef<Array<FocusActions | null>>([])
    const thisContentEditable = React.useRef<HTMLElement | null>(null);
    while (itemsRefArray.current.length < props.item.children.length) {
        itemsRefArray.current.push(null);
    }
    while (itemsRefArray.current.length > props.item.children.length) {
        itemsRefArray.current.pop();
    }

    const shouldUncollapse = shouldBeUncollapsed(props.item);
    const itemRef = React.useRef(props.item);
    React.useEffect(() => {
        // Force the focusedActionReceiver to look at the current item  otherwise it will fail to update
        itemRef.current = props.item
    });

    const focusThis = (end?: boolean) => {
        const currentElement = thisContentEditable.current;
        if (currentElement) {
            currentElement.focus();
            if (end) {
                const textNode = [...currentElement.childNodes].filter(i => i.nodeType == document.TEXT_NODE)[0];
                if (textNode && textNode.textContent) {
                    const range = document.createRange();
                    range.setStart(textNode, textNode.textContent.length)
                    range.setEnd(textNode, textNode.textContent.length)
                    const selection = window.getSelection();
                    if (selection) {
                        selection.removeAllRanges()
                        selection.addRange(range);
                    }
                }
            }
        }
        props.setThisAsFocused(focusedActionReceiver, { id: props.item.id, treePath: props.treePath });
    }

    const { show: showSideclipContextMenu, hideAll } = useContextMenu({
        id: SIDECLIP_CONTEXT_MENU_ID,
    });

    const raiseContextCopyIdEvent = (event: TriggerEvent) => {
        showSideclipContextMenu({ event });
        setTimeout(hideAll, 400);
        navigator.clipboard.writeText(props.item.id);
    };

    const jumpToSymlink = () => {
        if (props.item.symlinkedNode) {
            props.actions.focusItem({ id: props.item.symlinkedNode.id });
            return true;
        } else {
            return false;
        }
    }

    const focusedActionReceiver = makeFocusedActionReceiver({
        actions: props.actions,
        itemsRefArray,
        item: itemRef,
        raiseContextCopyIdEvent,
        jumpToSymlink,
        focusThis,
        treePath: props.treePath
    })

    props.pushRef(props.item.id, {
        focusThis,
        scrollThisIntoView: () => {
            thisContentEditable.current?.scrollIntoView()
        }
    })

    props.focusManager.current?.registerChild(props.treePath, props.item.id, {
        focus: focusThis
    })


    return <span className="itemWrapperClass">
        <EditableSection
            _ref={thisContentEditable}
            focusedActionReceiver={focusedActionReceiver}
            item={props.item}
            treePath={props.treePath}
            model={props.model}
            onFocusClick={focusThis}
            actions={props.actions}
            shouldUncollapse={shouldUncollapse}
            raiseContextCopyIdEvent={raiseContextCopyIdEvent}
            styleParams={props.styleParams}
        ></EditableSection>
        <ChildItems
            item={props.item}
            itemsRefArray={itemsRefArray}
            treePath={props.treePath}
            focusManager={props.focusManager}
            setThisAsFocused={props.setThisAsFocused}
            actions={props.actions}
            shouldUncollapse={shouldUncollapse}
            styleParams={props.styleParams}
            model={props.model}
            pushRef={props.pushRef}
        ></ChildItems>

    </span >
}

export type ItemRef = {
    focusThis: () => void,
    scrollThisIntoView: () => void
}

export default Item;
