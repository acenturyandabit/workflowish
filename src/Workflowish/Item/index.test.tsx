import * as renderer from 'react-test-renderer';
import * as React from "react";
import { it, expect } from '@jest/globals';
import Item from "."
import { makeItemActions } from '~Workflowish/mvc/controller';
import { ItemTreeNode } from '~Workflowish/mvc/model';
import { DFSFocusManager } from '~Workflowish/mvc/DFSFocus';

const mockEmptyFunction = () => {
    // Mock empty
}

const dummyItem: ItemTreeNode = {
    id: "",
    children: [],
    collapsed: false,
    lastModifiedUnixMillis: 0,
    searchHighlight: [],
    data: ""
};

it('Renders an item', () => {
    const model = {
        transformedData: {
            rootNode: dummyItem,
            keyedNodes: {},
            parentById: {}
        },
        setItemsByKey: mockEmptyFunction
    };
    const focusManager = new DFSFocusManager();
    const dummyAction = makeItemActions({
        treePath: [0],
        focusManager: focusManager,
        disableDelete: () => false,
        thisItem: dummyItem,
        model
    })
    const component = renderer.create(
        <Item
            item={{
                id: "test-id",
                lastModifiedUnixMillis: 0,
                data: "Hello world!",
                children: [],
                collapsed: true,
                searchHighlight: []
            }}
            treePath={[0]}
            focusManager={focusManager}
            pushRef={mockEmptyFunction}
            actions={dummyAction}
            setThisAsFocused={mockEmptyFunction}
            styleParams={{
                showId: false
            }}
            model={model}
        ></Item>
    )
    const item = component.toJSON();
    expect(item).toMatchSnapshot();
})