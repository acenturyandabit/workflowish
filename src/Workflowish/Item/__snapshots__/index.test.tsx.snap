// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders an item 1`] = `
<div
  style={
    {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
    }
  }
>
  <div
    style={
      {
        "flex": "1 0 auto",
        "margin": "0 5px 10px 5px",
      }
    }
  >
    <div
      className="search-bar"
    >
      <input
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="üîç Search"
        style={
          {
            "flex": "1 0 auto",
            "padding": "2px",
          }
        }
        value=""
      />
    </div>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0.2em",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚óè
          </span>
        </span>
        <div
          className="__editable___help_1"
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "Welcome to Workflowish!",
            }
          }
          data-testid="__help_1"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
      <div
        style={
          {
            "borderLeft": "1px solid white",
            "marginLeft": "0.5em",
            "paddingLeft": "5px",
          }
        }
      />
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñ∂
          </span>
        </span>
        <div
          className="__editable___help_2"
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "Workflowish is a keyboard-focused nested listing app.",
            }
          }
          data-testid="__help_2"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñº
          </span>
        </span>
        <div
          className="__editable___help_6"
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "This means you can nest bullet points...",
            }
          }
          data-testid="__help_6"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
      <div
        style={
          {
            "borderLeft": "1px solid white",
            "marginLeft": "0.5em",
            "paddingLeft": "5px",
          }
        }
      >
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0px",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚ñº
              </span>
            </span>
            <div
              className="__editable___help_7"
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "Like this,",
                }
              }
              data-testid="__help_7"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
          <div
            style={
              {
                "borderLeft": "1px solid white",
                "marginLeft": "0.5em",
                "paddingLeft": "5px",
              }
            }
          >
            <span
              className="itemWrapperClass"
            >
              <span
                onContextMenu={[Function]}
                style={
                  {
                    "background": "",
                  }
                }
              >
                <span
                  style={
                    {
                      "color": "white",
                      "cursor": "pointer",
                      "paddingLeft": "0px",
                      "userSelect": "none",
                    }
                  }
                >
                  <span
                    onClick={[Function]}
                  >
                    ‚ñº
                  </span>
                </span>
                <div
                  className="__editable___help_8"
                  contentEditable={true}
                  dangerouslySetInnerHTML={
                    {
                      "__html": "And this,",
                    }
                  }
                  data-testid="__help_8"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onInput={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  style={
                    {
                      "flex": "1 1 auto",
                      "scrollMarginTop": "100px",
                    }
                  }
                />
              </span>
              <div
                style={
                  {
                    "borderLeft": "1px solid white",
                    "marginLeft": "0.5em",
                    "paddingLeft": "5px",
                  }
                }
              >
                <span
                  className="itemWrapperClass"
                >
                  <span
                    onContextMenu={[Function]}
                    style={
                      {
                        "background": "",
                      }
                    }
                  >
                    <span
                      style={
                        {
                          "color": "white",
                          "cursor": "pointer",
                          "paddingLeft": "0.2em",
                          "userSelect": "none",
                        }
                      }
                    >
                      <span
                        onClick={[Function]}
                      >
                        ‚óè
                      </span>
                    </span>
                    <div
                      className="__editable___help_9"
                      contentEditable={true}
                      dangerouslySetInnerHTML={
                        {
                          "__html": "And so on.",
                        }
                      }
                      data-testid="__help_9"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onInput={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      style={
                        {
                          "flex": "1 1 auto",
                          "scrollMarginTop": "100px",
                        }
                      }
                    />
                  </span>
                  <div
                    style={
                      {
                        "borderLeft": "1px solid white",
                        "marginLeft": "0.5em",
                        "paddingLeft": "5px",
                      }
                    }
                  />
                </span>
              </div>
            </span>
          </div>
        </span>
      </div>
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñº
          </span>
        </span>
        <div
          className="__editable___help_10"
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "You can edit any item in the list by clicking it and typing :)",
            }
          }
          data-testid="__help_10"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
      <div
        style={
          {
            "borderLeft": "1px solid white",
            "marginLeft": "0.5em",
            "paddingLeft": "5px",
          }
        }
      >
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0.2em",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚óè
              </span>
            </span>
            <div
              className="__editable___help_11"
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "You can also add new items by pressing Enter on an existing item; or,",
                }
              }
              data-testid="__help_11"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
          <div
            style={
              {
                "borderLeft": "1px solid white",
                "marginLeft": "0.5em",
                "paddingLeft": "5px",
              }
            }
          />
        </span>
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0.2em",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚óè
              </span>
            </span>
            <div
              className="__editable___help_12"
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "Backspace all the way to the end, and then once more, to delete an item.",
                }
              }
              data-testid="__help_12"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
          <div
            style={
              {
                "borderLeft": "1px solid white",
                "marginLeft": "0.5em",
                "paddingLeft": "5px",
              }
            }
          />
        </span>
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0.2em",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚óè
              </span>
            </span>
            <div
              className="__editable___help_13"
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "Press CTRL+S to save your changes to your browser. (Other save options are available, read on!)",
                }
              }
              data-testid="__help_13"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
          <div
            style={
              {
                "borderLeft": "1px solid white",
                "marginLeft": "0.5em",
                "paddingLeft": "5px",
              }
            }
          />
        </span>
      </div>
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñ∂
          </span>
        </span>
        <div
          className="__editable___help_14"
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "Since lists get complicated, you can also collapse/show lists by focusing on an item and pressing CTRL+UP/DOWN.",
            }
          }
          data-testid="__help_14"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñº
          </span>
        </span>
        <div
          className="__editable___help_21"
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "Workflowish is inspired by Workflowy (https://workflowy.com/), but is free and open source, with in-browser storage or self-hosting.",
            }
          }
          data-testid="__help_21"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
      <div
        style={
          {
            "borderLeft": "1px solid white",
            "marginLeft": "0.5em",
            "paddingLeft": "5px",
          }
        }
      >
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0px",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚ñ∂
              </span>
            </span>
            <div
              className="__editable___help_22"
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "Workflowish also works on a mobile phone, again with a keyboard-first experience.",
                }
              }
              data-testid="__help_22"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
        </span>
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0.2em",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚óè
              </span>
            </span>
            <div
              className="__editable___help_26"
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "For self-hosted storage, clone the github repository (https://github.com/acenturyandabit/workflowish) and follow the README.",
                }
              }
              data-testid="__help_26"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
          <div
            style={
              {
                "borderLeft": "1px solid white",
                "marginLeft": "0.5em",
                "paddingLeft": "5px",
              }
            }
          />
        </span>
      </div>
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñ∂
          </span>
        </span>
        <div
          className="__editable___help_27"
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "As an extra treat for superusers out there, Workflowish also comes with:",
            }
          }
          data-testid="__help_27"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
    </span>
  </div>
</div>
`;
